services/device/geolocation/geoclue_location_provider.cc

```diff
@@ -0,0 +1,414 @@
+  if (!high_accuracy_requested_ && high_accuracy) {
+    if (client_state_ != kStopped) {
+      StopProvider();
+    }
+    high_accuracy_requested_ = true;
+  }
+
+  if (client_state_ != kStopped) {
+    return;
+  }
+  client_state_ = kInitializing;
+
+  dbus::ObjectProxy* proxy =
+      bus_->GetObjectProxy(kServiceName, dbus::ObjectPath(kManagerObjectPath));
+
+  dbus::MethodCall call(kManagerInterfaceName, "GetClient");
+  proxy->CallMethod(
+      &call, dbus::ObjectProxy::TIMEOUT_USE_DEFAULT,
+      base::BindOnce(&GeoClueLocationProvider::OnGetClientCompleted,
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void GeoClueLocationProvider::StopProvider() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  if (client_state_ == kStopped) {
+    return;
+  }
+
+  client_state_ = kStopped;
+
+  // Invalidate weak pointers, so we don't continue any async operations.
+  weak_ptr_factory_.InvalidateWeakPtrs();
+
+  // Stop can be called before the gclue_client_ has resolved.
+  if (!gclue_client_) {
+    return;
+  }
+
+  dbus::MethodCall stop(kClientInterfaceName, "Stop");
+  gclue_client_->CallMethod(&stop, dbus::ObjectProxy::TIMEOUT_USE_DEFAULT,
+                            base::DoNothing());
+
+  // Reset pointers to dbus objects.
+  gclue_client_.reset();
+}
+
+const mojom::Geoposition& GeoClueLocationProvider::GetPosition() {
+  return last_position_;
+}
+
+void GeoClueLocationProvider::OnPermissionGranted() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  permission_granted_ = true;
+  StartClient();
+}
+
+void GeoClueLocationProvider::SetPosition(const mojom::Geoposition& position) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  last_position_ = position;
+
+  if (client_state_ != kStarted) {
+    return;
+  }
+
+  if (last_position_.error_code == mojom::Geoposition_ErrorCode::NONE &&
+      !device::ValidateGeoposition(last_position_)) {
+    return;
+  }
+  position_update_callback_.Run(this, last_position_);
+}
+
+void GeoClueLocationProvider::OnGetClientCompleted(dbus::Response* response) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  if (!response) {
+    SetPosition(GetErrorPosition());
+    return;
+  }
+
+  dbus::MessageReader reader(response);
+  dbus::ObjectPath path;
+  if (!reader.PopObjectPath(&path)) {
+    SetPosition(GetErrorPosition());
+    return;
+  }
+
+  gclue_client_ = bus_->GetObjectProxy(kServiceName, path);
+  SetClientProperties();
+}
+
+void GeoClueLocationProvider::SetClientProperties() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  auto* raw_ptr = new GeoClueProperties(gclue_client_.get());
+  std::unique_ptr<dbus::PropertySet> properties(raw_ptr);
+
+  // Wait for all properties to be set before starting the client.
+  const auto callback = base::BarrierCallback<bool>(
+      2, base::BindOnce(&GeoClueLocationProvider::OnSetClientProperties,
+                        weak_ptr_factory_.GetWeakPtr(), std::move(properties)));
```

Comment:
why are we creating `properties` only to pass it to a callback on this class? Couldn't we just create the `GeoClueProperties` in `OnSetClientProperties`

Reply:
So I didn't want these to be owned by the `GeoClueLocationProvider` because it's lifetime isn't 1 to 1 with either of the `Dbus` interfaces which own these properties.

For example, on each `LocationUpdate` signal, a new `org.freedesktop.GeoClue2.Location` is created, at a new `ObjectPath`.

Similarly, a new `org.freedesktop.GeoClue2.Client` may be created after stopping and starting the `GeoClueLocationProvider`.

I thought it might be less error prone if I only kept the properties around as long as they're being used:
1. `GeoClueProperties` just long enough to set the properties we care about.
2. `GeoClueLocationProperties` just long enough to read the location data.

I'm happy to convert these to members on `GeoClueLocationProvider` if you think it'll make things more obvious.

Reply:
Adding a `::Properties` structs :+1: (https://github.com/brave/brave-core/pull/18063#discussion_r1196552245 has more context)